File: vertexai/app/src/main/java/com/google/firebase/example/vertexai/java/GenerateContentViewModel.java
package com.google.firebase.example.vertexai.java;

import android.content.ContentResolver;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;

import androidx.lifecycle.ViewModel;

import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.firebase.example.vertexai.R;
import com.google.firebase.vertexai.FirebaseVertexAI;
import com.google.firebase.vertexai.GenerativeModel;
import com.google.firebase.vertexai.java.GenerativeModelFutures;
import com.google.firebase.vertexai.type.Content;
import com.google.firebase.vertexai.type.CountTokensResponse;
import com.google.firebase.vertexai.type.GenerateContentResponse;

import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.concurrent.Executor;

public class GenerateContentViewModel extends ViewModel {
    private GenerativeModelFutures model;

    // Only meant to separate the scope of the initialization snippet
    // so that it doesn't cause a naming clash with the top level declaration
    static class InitializationSnippet {
        // [START initialize_model]
        GenerativeModel gm = FirebaseVertexAI.getInstance()
                .generativeModel("gemini-1.5-pro-preview-0409");

        GenerativeModelFutures model = GenerativeModelFutures.from(gm);
        // [END initialize_model]
    }

    void generateContentStream() {
        // [START text_gen_text_only_prompt_streaming]
        Content prompt = new Content.Builder()
                .addText("Write a story about a magic backpack.")
                .build();

        Publisher<GenerateContentResponse> streamingResponse =
                model.generateContentStream(prompt);

        final String[] fullResponse = {""};

        streamingResponse.subscribe(new Subscriber<GenerateContentResponse>() {
            @Override
            public void onNext(GenerateContentResponse generateContentResponse) {
                String chunk = generateContentResponse.getText();
                fullResponse[0] += chunk;
            }

            @Override
            public void onComplete() {
                System.out.println(fullResponse[0]);
            }

            @Override
            public void onError(Throwable t) {
                t.printStackTrace();
            }

            @Override
            public void onSubscribe(Subscription s) {
            }
        });
        // [END text_gen_text_only_prompt_streaming]
    }

    void generateContent(Executor executor) {
        // [START text_gen_text_only_prompt]
        // Provide a prompt that contains text
        Content prompt = new Content.Builder()
                .addText("Write a story about a magic backpack.")
                .build();

        // To generate text output, call generateContent with the text input
        ListenableFuture<GenerateContentResponse> response = model.generateContent(prompt);
        Futures.addCallback(response, new FutureCallback<GenerateContentResponse>() {
            @Override
            public void onSuccess(GenerateContentResponse result) {
                String resultText = result.getText();
                System.out.println(resultText);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        }, executor);
        // [END text_gen_text_only_prompt]
    }

    // Fake implementation to exemplify Activity.getResources()
    Resources getResources() {
        return null;
    }

    // Fake implementation to exemplify Activity.getApplicationContext()
    Context getApplicationContext() {
        return null;
    }

    void generateContentWithImageStream() {
        // [START text_gen_multimodal_one_image_prompt_streaming]
        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.sparky);

        Content prompt = new Content.Builder()
                .addImage(bitmap)
                .addText("Write a story about a magic backpack.")
                .build();

        Publisher<GenerateContentResponse> streamingResponse =
                model.generateContentStream(prompt);

        final String[] fullResponse = {""};

        streamingResponse.subscribe(new Subscriber<GenerateContentResponse>() {
            @Override
            public void onNext(GenerateContentResponse generateContentResponse) {
                String chunk = generateContentResponse.getText();
                fullResponse[0] += chunk;
            }

            @Override
            public void onComplete() {
                System.out.println(fullResponse[0]);
            }

            @Override
            public void onError(Throwable t) {
                t.printStackTrace();
            }

            @Override
            public void onSubscribe(Subscription s) {
            }
        });
        // [END text_gen_multimodal_one_image_prompt_streaming]
    }

    void generateContentWithImage(Executor executor) {
        // [START text_gen_multimodal_one_image_prompt]
        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.sparky);

        Content content = new Content.Builder()
                .addImage(bitmap)
                .addText("What developer tool is this mascot from?")
                .build();

        ListenableFuture<GenerateContentResponse> response = model.generateContent(content);
        Futures.addCallback(response, new FutureCallback<GenerateContentResponse>() {
            @Override
            public void onSuccess(GenerateContentResponse result) {
                String resultText = result.getText();
                System.out.println(resultText);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        }, executor);
        // [END text_gen_multimodal_one_image_prompt]
    }

    void generateContentWithMultipleImagesStream() {
        // [START text_gen_multimodal_multi_image_prompt_streaming]
        Bitmap bitmap1 = BitmapFactory.decodeResource(getResources(), R.drawable.sparky);
        Bitmap bitmap2 = BitmapFactory.decodeResource(getResources(), R.drawable.sparky_eats_pizza);

        // Provide a prompt that includes the images specifed above and text
        Content prompt = new Content.Builder()
                .addImage(bitmap1)
                .addImage(bitmap2)
                .addText("What's different between these pictures?")
                .build();

        Publisher<GenerateContentResponse> streamingResponse =
                model.generateContentStream(prompt);

        final String[] fullResponse = {""};

        streamingResponse.subscribe(new Subscriber<GenerateContentResponse>() {
            @Override
            public void onNext(GenerateContentResponse generateContentResponse) {
                String chunk = generateContentResponse.getText();
                fullResponse[0] += chunk;
            }

            @Override
            public void onComplete() {
                System.out.println(fullResponse[0]);
            }

            @Override
            public void onError(Throwable t) {
                t.printStackTrace();
            }

            @Override
            public void onSubscribe(Subscription s) {
            }
        });
        // [END text_gen_multimodal_multi_image_prompt_streaming]
    }

    void generateContentWithMultipleImages(Executor executor) {
        // [START text_gen_multimodal_multi_image_prompt]
        Bitmap bitmap1 = BitmapFactory.decodeResource(getResources(), R.drawable.sparky);
        Bitmap bitmap2 = BitmapFactory.decodeResource(getResources(), R.drawable.sparky_eats_pizza);

        // Provide a prompt that includes the images specifed above and text
        Content prompt = new Content.Builder()
                .addImage(bitmap1)
                .addImage(bitmap2)
                .addText("What's different between these pictures?")
                .build();

        ListenableFuture<GenerateContentResponse> response = model.generateContent(prompt);
        Futures.addCallback(response, new FutureCallback<GenerateContentResponse>() {
            @Override
            public void onSuccess(GenerateContentResponse result) {
                String resultText = result.getText();
                System.out.println(resultText);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        }, executor);
        // [END text_gen_multimodal_multi_image_prompt]
    }

    void generateContentWithVideo(Executor executor, Uri videoUri) {
        // [START text_gen_multimodal_video_prompt]
        ContentResolver resolver = getApplicationContext().getContentResolver();
        try (InputStream stream = resolver.openInputStream(videoUri)) {
            File videoFile = new File(new URI(videoUri.toString()));
            int videoSize = (int) videoFile.length();
            byte[] videoBytes = new byte[videoSize];
            if (stream != null) {
                stream.read(videoBytes, 0, videoBytes.length);
                stream.close();

                Content prompt = new Content.Builder()
                        .addBlob("video/mp4", videoBytes)
                        .addText("What is in the video?")
                        .build();

                ListenableFuture<GenerateContentResponse> response = model.generateContent(prompt);
                Futures.addCallback(response, new FutureCallback<GenerateContentResponse>() {
                    @Override
                    public void onSuccess(GenerateContentResponse result) {
                        String resultText = result.getText();
                        System.out.println(resultText);
                    }

                    @Override
                    public void onFailure(Throwable t) {
                        t.printStackTrace();
                    }
                }, executor);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        // [END text_gen_multimodal_video_prompt]
    }

    void generateContentWithVideoStream(
            Uri videoUri
    ) {
        // [START text_gen_multimodal_video_prompt_streaming]
        ContentResolver resolver = getApplicationContext().getContentResolver();
        try (InputStream stream = resolver.openInputStream(videoUri)) {
            File videoFile = new File(new URI(videoUri.toString()));
            int videoSize = (int) videoFile.length();
            byte[] videoBytes = new byte[videoSize];
            if (stream != null) {
                stream.read(videoBytes, 0, videoBytes.length);
                stream.close();

                Content prompt = new Content.Builder()
                        .addBlob("video/mp4", videoBytes)
                        .addText("What is in the video?")
                        .build();

                Publisher<GenerateContentResponse> streamingResponse =
                        model.generateContentStream(prompt);

                final String[] fullResponse = {""};

                streamingResponse.subscribe(new Subscriber<GenerateContentResponse>() {
                    @Override
                    public void onNext(GenerateContentResponse generateContentResponse) {
                        String chunk = generateContentResponse.getText();
                        fullResponse[0] += chunk;
                    }

                    @Override
                    public void onComplete() {
                        System.out.println(fullResponse[0]);
                    }

                    @Override
                    public void onError(Throwable t) {
                        t.printStackTrace();
                    }

                    @Override
                    public void onSubscribe(Subscription s) {
                    }
                });
            }
        } catch (IOException e) {
            e.printStackTrace();
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        // [END text_gen_multimodal_video_prompt_streaming]
    }

    void countTokensText(Executor executor) {
        // [START count_tokens_text]
        Content text = new Content.Builder()
                .addText("Write a story about a magic backpack.")
                .build();

        ListenableFuture<CountTokensResponse> countTokensResponse = model.countTokens(text);

        Futures.addCallback(countTokensResponse, new FutureCallback<CountTokensResponse>() {
            @Override
            public void onSuccess(CountTokensResponse result) {
                int totalTokens = result.getTotalTokens();
                int totalBillableTokens = result.getTotalBillableCharacters();
                System.out.println("totalTokens = " + totalTokens +
                        "totalBillableTokens = " + totalBillableTokens);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        }, executor);
        // [END count_tokens_text]
    }

    void countTokensMultimodal(Executor executor, Bitmap bitmap) {
        // [START count_tokens_text_image]
        Content text = new Content.Builder()
                .addImage(bitmap)
                .addText("Where can I buy this")
                .build();

        // For text-only input
        ListenableFuture<CountTokensResponse> countTokensResponse = model.countTokens(text);

        Futures.addCallback(countTokensResponse, new FutureCallback<CountTokensResponse>() {
            @Override
            public void onSuccess(CountTokensResponse result) {
                int totalTokens = result.getTotalTokens();
                int totalBillableTokens = result.getTotalBillableCharacters();
                System.out.println("totalTokens = " + totalTokens +
                        "totalBillableTokens = " + totalBillableTokens);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        }, executor);
        // [END count_tokens_text_image]
    }
}

File: vertexai/app/src/main/java/com/google/firebase/example/vertexai/kotlin/GenerateContentViewModel.kt
package com.google.firebase.example.vertexai.kotlin

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.Firebase
import com.google.firebase.example.vertexai.R
import com.google.firebase.vertexai.GenerativeModel
import com.google.firebase.vertexai.type.content
import com.google.firebase.vertexai.vertexAI
import kotlinx.coroutines.launch

@Suppress("JoinDeclarationAndAssignment") // for the generativeModel var
class GenerateContentViewModel : ViewModel() {
    private val TAG = "ContentViewModel"
    private var generativeModel: GenerativeModel

    // Only meant to separate the scope of the initialization snippet
    // so that it doesn't cause a naming clash with the top level generativeModel
    fun initialize() {
        // [START initialize_model]
        val generativeModel = Firebase.vertexAI.generativeModel(
            // Specify a model that supports your use case
            // Gemini 1.5 Pro is versatile and can accept both text-only and multimodal prompt inputs
            modelName = "gemini-1.5-pro-preview-0409"
        )
        // [END initialize_model]
    }

    init {
        generativeModel = Firebase.vertexAI.generativeModel("gemini-1.5-pro-preview-0409")
    }

    fun generateContentStream() {
        viewModelScope.launch {
            // [START text_gen_text_only_prompt_streaming]
            // Provide a prompt that includes only text
            val prompt = "Write a story about a magic backpack."
            // To stream generated text output, call generateContentStream and pass in the prompt
            var fullResponse = ""
            generativeModel.generateContentStream(prompt).collect { chunk ->
                Log.d(TAG, chunk.text ?: "")
                fullResponse += chunk.text
            }
            // [END text_gen_text_only_prompt_streaming]
        }
    }

    fun generateContent() {
        viewModelScope.launch {
            // [START text_gen_text_only_prompt]
            // Provide a prompt that includes only text
            val prompt = "Write a story about a magic backpack."

            // To generate text output, call generateContent and pass in the prompt
            val response = generativeModel.generateContent(prompt)
            Log.d(TAG, response.text ?: "")
            // [END text_gen_text_only_prompt]
        }
    }

    fun generateContentWithImageStream(resources: Resources) {
        viewModelScope.launch {
            // [START text_gen_multimodal_one_image_prompt_streaming]
            // Loads an image from the app/res/drawable/ directory
            val bitmap: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.sparky)

            val prompt = content {
                image(bitmap)
                text("What developer tool is this mascot from?")
            }

            var fullResponse = ""
            generativeModel.generateContentStream(prompt).collect { chunk ->
                Log.d(TAG, chunk.text ?: "")
                fullResponse += chunk.text
            }
            // [END text_gen_multimodal_one_image_prompt_streaming]
        }
    }

    fun generateContentWithImage(resources: Resources) {
        viewModelScope.launch {
            // [START text_gen_multimodal_one_image_prompt]
            // Loads an image from the app/res/drawable/ directory
            val bitmap: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.sparky)

            val prompt = content {
                image(bitmap)
                text("What developer tool is this mascot from?")
            }

            val response = generativeModel.generateContent(prompt)
            Log.d(TAG, response.text ?: "")
            // [END text_gen_multimodal_one_image_prompt]
        }
    }

    fun generateContentWithMultipleImagesStream(resources: Resources) {
        viewModelScope.launch {
            // [START text_gen_multimodal_multi_image_prompt_streaming]
            // Loads an image from the app/res/drawable/ directory
            val bitmap1: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.sparky)
            val bitmap2: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.sparky_eats_pizza)

            val prompt = content {
                image(bitmap1)
                image(bitmap2)
                text("What is different between these pictures?")
            }

            var fullResponse = ""
            generativeModel.generateContentStream(prompt).collect { chunk ->
                Log.d(TAG, chunk.text ?: "")
                fullResponse += chunk.text
            }
            // [END text_gen_multimodal_multi_image_prompt_streaming]
        }
    }

    fun generateContentWithMultipleImages(resources: Resources) {
        viewModelScope.launch {
            // [START text_gen_multimodal_multi_image_prompt]
            // Loads an image from the app/res/drawable/ directory
            val bitmap1: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.sparky)
            val bitmap2: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.sparky_eats_pizza)

            val prompt = content {
                image(bitmap1)
                image(bitmap2)
                text("What is different between these pictures?")
            }

            val response = generativeModel.generateContent(prompt)
            Log.d(TAG, response.text ?: "")
            // [END text_gen_multimodal_multi_image_prompt]
        }
    }

    fun generateContentWithVideoStream(
        applicationContext: Context,
        videoUri: Uri
    ) {
        viewModelScope.launch {
            // [START text_gen_multimodal_video_prompt_streaming]
            val contentResolver = applicationContext.contentResolver
            contentResolver.openInputStream(videoUri).use { stream ->
                stream?.let {
                    val bytes = stream.readBytes()

                    val prompt = content {
                        blob("video/mp4", bytes)
                        text("What is in the video?")
                    }

                    var fullResponse = ""
                    generativeModel.generateContentStream(prompt).collect { chunk ->
                        Log.d(TAG, chunk.text ?: "")
                        fullResponse += chunk.text
                    }
                }
            }
            // [END text_gen_multimodal_video_prompt_streaming]
        }
    }

    fun generateContentWithVideo(
        applicationContext: Context,
        videoUri: Uri
    ) {
        viewModelScope.launch {
            // [START text_gen_multimodal_video_prompt]
            val contentResolver = applicationContext.contentResolver
            contentResolver.openInputStream(videoUri).use { stream ->
                stream?.let {
                    val bytes = stream.readBytes()

                    val prompt = content {
                        blob("video/mp4", bytes)
                        text("What is in the video?")
                    }

                    val response = generativeModel.generateContent(prompt)
                    Log.d(TAG, response.text ?: "")
                }
            }
            // [END text_gen_multimodal_video_prompt]
        }
    }

    fun countTokensText() {
        viewModelScope.launch {
            // [START count_tokens_text]
            val (tokens, billableChars) = generativeModel.countTokens("Write a story about a magic backpack.")
            // [END count_tokens_text]
        }
    }

    fun countTokensMultimodal(bitmap: Bitmap) {
        viewModelScope.launch {
            // [START count_tokens_text_image]
            val prompt = content {
                image(bitmap)
                text("Where can I buy this?")
            }
            val (tokens, billableChars) = generativeModel.countTokens(prompt)
            // [END count_tokens_text_image]
        }
    }
}